libs/Microsoft.MixedReality.WebRTC.Unity/Assets/Microsoft.MixedReality.WebRTC.Unity/Scripts/Media/RemoteAudioSource.cs

```diff
@@ -40,6 +40,163 @@ public class RemoteAudioSource : AudioSource
+
+        /// <summary>
+        /// Set when we've enqueued the job to create the audio clip on the main thread. (to avoid duplication)
+        /// </summary>
+        private volatile bool _audioClipEnqueuedCreate = false;
+
+        /// <summary>
+        /// Internal adaptor from webrtc frame data to unity audio clip.
+        /// </summary>
+        class AudioData
+        {
+            float[] _buffer = null;
+            bool _isFull = false;
+            int _readPos = 0; // AudioClip fetches start from here
+            int _writePos = 0; // WebRTC data gets appended here
+            int _channelCount = 0;
+            int _sampleRate = 0;
+            public Action<int, int> AudioFormatChanged = null;
+
+            void UpdateInternalFormat(AudioFrame frame)
+            {
+                if (frame.sampleRate != _sampleRate
+                    || frame.channelCount != _channelCount
+                    || _buffer == null
+                    || _buffer.Length != (frame.sampleRate * frame.channelCount))
+                {
+                    // If anything changes, let's drop the buffer for simplicity (i.e. we won't resample or do extra work)
+                    _sampleRate = (int)frame.sampleRate;
+                    _channelCount = (int)frame.channelCount;
+                    _buffer = new float[frame.sampleRate * frame.channelCount];
+                    _readPos = 0;
+                    _writePos = 0;
+                    AudioFormatChanged?.Invoke(_sampleRate, _channelCount);
+                }
+            }
+
+            internal void AppendAudio16(AudioFrame frame)
+            {
+                Debug.Assert(Monitor.IsEntered(this));
+                UpdateInternalFormat(frame);
+                Debug.Assert((frame.frameCount * frame.channelCount) <= _buffer.Length);
+                unsafe
+                {
+                    short* src = (short*)frame.audioData;
+                    int spaceAtEnd = _buffer.Length - _writePos;
+                    int srcSamples = (int)(frame.frameCount * frame.channelCount);
+                    // write1 is from _writePos forwards. write2 is the wraparound part (very often 0)
+                    int write1 = Math.Min(spaceAtEnd, srcSamples);
+                    for (int i = 0; i < write1; ++i)
+                    {
+                        _buffer[i + _writePos] = src[i] / 32768.0f; // 16 bit data is signed. Remap to -1,1
+                    }
+
+                    // Adjust the _writePos. Note that this may also update *_readPos*, if the reader is slow.
+                    // We will always enque ALL the data on the basis that newer is better.
+                    // Sometimes, if the reader isn't consuming fast enough, we may advance 'through' the read position.
+                    // This will give bad artifacts, so instead we'll drop some of the input instead.
```

Comment:
Isn't that comment directly contradicting the one just above? Which is it, we discard new input or overwrite old non-consumed one?